File: leetcode/java/FindTheDuplicateNumber.java
class FindTheDuplicateNumber {
    public int findDuplicate(int[] nums){
        int hare = 0, tortoise = 0;
        while(tortoise<nums.length){
            if(nums[hare]==nums[tortoise] && tortoise!=hare){
                return nums[hare];
            }
            hare+=2;
            hare = hare % nums.length;
            tortoise+=1;
        }
        return -1;
    }
}

File: leetcode/cpp/find-the-duplicate-number.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution{
    public:
    int findDuplicate(vector<int> &nums){
        for(int i = 0; i < nums.size(); i++){
            int idx = abs(nums[i]);
            if(nums[idx] < 0){
                return idx;
            }else{
                nums[idx]=-nums[idx];
            }
        }
        return -1;
    }
};